Add SparseObservable support to evolved_operator_ansatz

akhil-40409 opened the pull request:
## PR Summary: Add SparseObservable Support to Evolved Operator Ansatz 

### Overview
---
Adds support for SparseObservable in evolved_operator_ansatz, enabling the Rust-accelerated path for better performance.


### Changes
---
evolved_operator_ansatz.py:
- Added conditional import and handling for SparseObservable
- Extended Rust fast path to accept SparseObservable (in addition to SparsePauliOp)
- Fixed _is_pauli_identity() to handle SparseObservable with coefficient checks 
- Fixed edge cases: return empty circuit when all operators are filtered out, handle string/list parameter_prefix in _remove_identities()

test_evolved_op_ansatz.py:
- Added tests for SparseObservable acceptance and equivalence
- New test class TestEvolvedOperatorAnsatzSparseObservable covering Rust path, identity removal, mixed operators, and equivalence with SparsePauliOp

akhil-40409 committed:
Add SparseObservable support to evolved_operator_ansatz

- Add conditional import and handling for SparseObservable
- Extend Rust fast path to accept SparseObservable (in addition to SparsePauliOp)
- Enhance _is_pauli_identity() to handle SparseObservable with coefficient checks
- Fix edge cases: return empty circuit when all operators filtered, handle string/list parameter_prefix
- Add comprehensive test coverage for SparseObservable support

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=15442) <br/>All committers have signed the CLA.

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -36,6 +36,11 @@
 if typing.TYPE_CHECKING:
     from qiskit.synthesis.evolution import EvolutionSynthesis
 
+try:
```
In which case can this be None? There should never be an import error here

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -106,10 +114,14 @@ def evolved_operator_ansatz(
         operators, parameter_prefix = _remove_identities(operators, parameter_prefix)
 
+    # After removing identities, check if we have any operators left
+    if len(operators) == 0:
```
Evolving an identity operator corresponds to adding a global phase, this is not an empty circuit.

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -498,16 +519,97 @@ def _validate_prefix(parameter_prefix, operators):
+#     if isinstance(operator, Pauli):
+#         return not np.any(np.logical_or(operator.x, operator.z))
+#     return False
+def _is_pauli_identity(operator, tol=1e-10):
```
We can simply check if a `SparseObservable` is the identity by doing
```python
obs = # some SparseObservable
is_identity = obs.num_terms == 1 && len(obs[0].bit_labels()) == 0
```
Since the observable must have exactly 1 term and the bit labels must be empty (since identities don't have labels).

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -498,16 +519,97 @@ def _validate_prefix(parameter_prefix, operators):
     if isinstance(operator, Pauli):
-        return not np.any(np.logical_or(operator.x, operator.z))
-    return False
+        # Pauli identity if no X or Z bits are set
```
This shouldn't be changed

Cryoris reviewed (commented):
This PR is over-engineering the solution a lot, please check the comments below to simplify the code.

akhil-40409 committed:
Fix import after removing try

akhil-40409 committed:
Address PR review comments for SparseObservable support

- Use direct import of SparseObservable (no try-except)
- Simplify identity check: use num_terms == 1 and len(bit_labels()) == 0
- Fix empty circuit handling: return circuit with correct num_qubits when all operators are identities
- Add SparseObservable support to fast Rust path
- Fix prefix handling to support both string and list types
- Add comprehensive test cases for all changes

akhil-40409 commented:
Hi, I’ve updated and simplified the code according to your comments.

Cryoris commented on test/python/circuit/library/test_evolved_op_ansatz.py:
```diff
@@ -205,6 +205,186 @@ def test_sympify_is_real(self):
+            self.SparseObservable = SparseObservable
+        except ImportError:
+            self.SparseObservable = None
+
```
When running the test locally, do you need this try-except for the import? If `qiskit` is installed we shouldn't need this and just have a global 
```python
from qiskit.quantum_info import SparseObservable
```
import

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -110,7 +113,10 @@ def evolved_operator_ansatz(
         operators, parameter_prefix = _remove_identities(operators, parameter_prefix)
 
-    if any(op.num_qubits != num_qubits for op in operators):
+    # After removing identities, update num_operators to reflect the actual count
```
We could optimize this a bit: first, we can change to update `num_operators` only inside the `if remove_identities` statement (otherwise the number of operators won't change). Then, if `num_operators` is in fact 0 after the update, we can directly return an empty circuit of size `num_qubits` (also still inside the if-statement). With this we are sure that `num_operators > 0` in the rest of the code and we don't have to check that anymore.

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -127,9 +133,13 @@ def evolved_operator_ansatz(
         and evolution is None
-        and all(isinstance(op, SparsePauliOp) for op in operators)
+        and all(
+            isinstance(op, SparsePauliOp) or isinstance(op, SparseObservable)
```
You can merge two isinstance checks as 
```suggestion
            isinstance(op, (SparsePauliOp, SparseObservable))
```
(this also occurs somewhere else in the code)

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -516,6 +528,12 @@ def _remove_identities(operators, prefixes):
 
     cleaned_ops = [op for i, op in enumerate(operators) if i not in identity_ops]
-    cleaned_prefix = [prefix for i, prefix in enumerate(prefixes) if i not in identity_ops]
+    # Handle both string and list prefixes
```
Nice catch! This fixes a bug we had before. It would be nice to backport this fix to the 2.3 series. Would you be up for splitting this fix into a separate PR (including a test) that we merge separately from this? If not no worries, I can do a small follow up to fix it in 2.3.

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -516,6 +528,12 @@ def _remove_identities(operators, prefixes):
+        cleaned_prefix = [prefix for i, prefix in enumerate(prefixes) if i not in identity_ops]
 
-    return cleaned_ops, cleaned_prefix
+    return cleaned_ops, cleaned_prefix
```
Could we leave the final empty lines in place? 🙂

Cryoris commented on test/python/circuit/library/test_evolved_op_ansatz.py:
```diff
@@ -205,6 +205,186 @@ def test_sympify_is_real(self):
+        self.assertTrue(all("theta" in name for name in param_names))
+
+
+class TestEvolvedOperatorAnsatzSparseObservable(QiskitTestCase):
```
We don't really need a new test class for this, could you just add the functions to the existing test class above?

Cryoris replied:
@akhil-102 this comment is also still open

akhil-40409 replied:
Fixed it

Cryoris commented on test/python/circuit/library/test_evolved_op_ansatz.py:
```diff
@@ -205,6 +205,186 @@ def test_sympify_is_real(self):
+        self.assertEqual(ansatz.num_qubits, 1)
+        self.assertEqual(ansatz.num_parameters, 1)
+
+    def test_sparse_observable_identity_detection(self):
```
I think we can remove this test, we don't usually test internal, private functions since they are subject to change and instead focus on testing the user API. Since you have other tests that check identities are being removed, I would say it's safe to say the internal function works

Cryoris reviewed (commented):
Thanks for the updates @akhil-102, the approach looks good now! Also great catch on the bug with the string-based parameter prefix. It would be great to separate the fix into a different PR, if you'd be interested in doing that -- see also the comments below.

akhil-40409 committed:
Optimize evolved_operator_ansatz and clean up tests

- Move num_operators update inside remove_identities block for optimization
- Return early with empty circuit when all operators are identities
- Remove num_operators > 0 checks since we guarantee it after identity removal
- Replace try-except import pattern with direct SparseObservable import in tests
- Remove test for private _is_pauli_identity function
- Update all test references to use SparseObservable directly

akhil-40409 commented:
Hey
I’ve updated the code based on your comments, except for the bug fix PR.

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -93,7 +93,10 @@ def evolved_operator_ansatz(
+    # Check for SparseObservable first (it's iterable, so isinstance check must come before len check)
+    if isinstance(operators, SparseObservable):
+        operators = [operators]
+    elif isinstance(operators, BaseOperator):
```
You can just merge these instance checks, as
```suggestion
    if isinstance(operators, (BaseOperator, SparseObservable)):
```

Cryoris commented on qiskit/circuit/library/n_local/evolved_operator_ansatz.py:
```diff
@@ -169,7 +179,7 @@ def evolved_operator_ansatz(
                 flatten_operator = False
 
-            elif isinstance(op, BaseOperator):
+            elif isinstance(op, BaseOperator) or isinstance(op, SparseObservable):
```
```suggestion
            elif isinstance(op, (BaseOperator, SparseObservable)):
```

akhil-40409 committed:
Update qiskit/circuit/library/n_local/evolved_operator_ansatz.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

akhil-40409 committed:
Update qiskit/circuit/library/n_local/evolved_operator_ansatz.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

akhil-40409 committed:
Combined SparseObservable tests into TestEvolvedOperatorAnsatz

Cryoris commented:
Thanks for the fixes! Could you also add a bugfix release note, since you're now fixing a bug in this PR?

akhil-40409 committed:
Add release note for SparseObservable support in evolved_operator_ansatz

akhil-40409 commented:
Added a bugfix release note as requested.

Cryoris commented on releasenotes/notes/fix-evolved-operator-ansatz-sparseobservable-ddcabb3853ca5fd5.yaml:
```diff
@@ -0,0 +1,9 @@
+    support :class:`~qiskit.quantum_info.SparseObservable` operators. The function now
+    correctly accepts ``SparseObservable`` instances and uses the fast Rust path when
+    ``flatten=True`` and no custom evolution is specified, matching the behavior for
+    :class:`~qiskit.quantum_info.SparsePauliOp`.
```
What you're describing here is a _new feature_, the bug fix is around the parameter prefix. 
```suggestion
features_circuits:
  - |
    Added support for :class:`.SparseObservable` to 
    :func:`~qiskit.circuit.library.evolved_operator_ansatz`.
fixes:
  - |
    Fixed a bug in :func:`~qiskit.circuit.library.evolved_operator_ansatz`, where
    setting the ``parameter_prefix`` argument to a single string could cause an 
    error if identity terms were removed, depending on the order of operators.
```

Cryoris committed:
Update reno

akhil-40409 committed:
Apply black formatting

akhil-40409 commented:
There was a formatting error with the merge; I fixed it.

akhil-40409 commented:
All CI checks are passing. Code owner review requested.

Cryoris reviewed (approved):
Thanks for the contribution @akhil-102 !
